atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java

```diff
@@ -64,44 +66,67 @@ public static Workflow create(
-        return DefaultWorkflow.create(
-                store,
-                (txn, taskIndex) -> run(store, workflowConfiguration, taskIndex),
-                workflowConfiguration,
-                executionExecutor);
+        TransientRowsTask transientRowsTask =
+                new TransientRowsTask(workflowConfiguration, TransientRowsWorkflows::logViolations);
+        return DefaultWorkflow.create(store, transientRowsTask, workflowConfiguration, executionExecutor);
     }
 
-    private static Optional<WitnessedTransaction> run(
-            InteractiveTransactionStore store, TransientRowsWorkflowConfiguration configuration, int taskIndex) {
-        return store.readWrite(txn -> {
-            String tableName = configuration.tableConfiguration().tableName();
-            txn.write(tableName, ImmutableWorkloadCell.of(SUMMARY_ROW, taskIndex), VALUE);
-            txn.write(tableName, ImmutableWorkloadCell.of(taskIndex, COLUMN), VALUE);
-            if (taskIndex > 0) {
-                int indexToRead = SECURE_RANDOM.nextInt(taskIndex);
-                txn.read(tableName, ImmutableWorkloadCell.of(SUMMARY_ROW, indexToRead));
-                txn.read(tableName, ImmutableWorkloadCell.of(indexToRead, COLUMN));
-
-                if (txn.read(tableName, ImmutableWorkloadCell.of(taskIndex - 1, COLUMN))
-                        .isPresent()) {
-                    txn.delete(tableName, ImmutableWorkloadCell.of(SUMMARY_ROW, taskIndex - 1));
-                    txn.delete(tableName, ImmutableWorkloadCell.of(taskIndex - 1, COLUMN));
+    public static Workflow create(
+            InteractiveTransactionStore store,
+            TransientRowsWorkflowConfiguration workflowConfiguration,
+            ListeningExecutorService executionExecutor,
+            Consumer<List<CrossCellInconsistency>> onFailure) {
+        TransientRowsTask transientRowsTask = new TransientRowsTask(workflowConfiguration, onFailure);
+        return DefaultWorkflow.create(store, transientRowsTask, workflowConfiguration, executionExecutor);
+    }
+
+    private static final class TransientRowsTask
+            extends StoppableKeyedTransactionTask<InteractiveTransactionStore, List<CrossCellInconsistency>> {
+
+        private final TransientRowsWorkflowConfiguration configuration;
+
+        private TransientRowsTask(
+                TransientRowsWorkflowConfiguration configuration, Consumer<List<CrossCellInconsistency>> onFailure) {
+            super(new AtomicBoolean(), onFailure);
+            this.configuration = configuration;
+        }
+
+        @Override
+        protected Optional<WitnessedTransaction> run(InteractiveTransactionStore store, Integer taskIndex) {
+            if (taskIndex % configuration.validateEveryNIterations() == 0) {
+                // This is cheeky, as we are not including our reads in our transaction history!
+                List<CrossCellInconsistency> violations = findInconsistencyInFinalIndexState(configuration, store);
```

Comment:
Even though we "recordFailure" and then stop, there's no guarantee that the invariant reporter would also catch this. I think for the vast majority of cases it would.